File: code/ch06/6.6.1.split-array-largest-sum.java
class Solution {
    public int splitArray(int[] nums, int m) {
        long maxNum = nums[0], sumNum = 0;
        for (long num : nums) {
            maxNum = Math.max(maxNum, num);
            sumNum += num;
        }
        long l = maxNum, h = sumNum;
        while (l <= h) {
            long mid = l + (h - l) / 2;
            if (l == h) {
                return (int) l;
            } else if (help(nums, mid, m)) {
                h = mid;
            } else {
                l = mid + 1;
            }
        }
        return -1;
    }

    public boolean help(int[] nums, long ans, int m) {
        long cnt = 1, cur = 0;
        for (long num : nums) {
            if (cur + num > ans) {
                cur = num;
                cnt += 1;
            } else {
                cur += num;
            }
        }
        return cnt <= m;
    }
}

File: code/ch06/6.6.1.split-array-largest-sum.py
class Solution:
    def splitArray(self, nums: List[int], m: int) -> int:
        # 判断 ans 是否可行
        def help(ans: int) -> boolean:
            cnt, cur = 1, 0
            for num in nums:
                if (cur + num) > ans:
                    cur = num
                    cnt += 1
                else:
                    cur += num
            return cnt <= m

        l, h = max(nums), sum(nums)
        while l <= h:
            mid = l + (h - l) // 2
            if l == h:
                return l
            elif help(mid):
                h = mid
            else:
                l = mid + 1
        return -1
